src/rgw/rgw_lc.cc

```diff
@@ -1249,32 +1252,244 @@ class LCOpAction_Transition : public LCOpAction {
+    obj_op.meta.delete_at = real_time();
+    bufferlist blo;
+    blo.append("");
+    obj_op.meta.data = &blo;
+    obj_op.meta.if_match = NULL;
+    obj_op.meta.if_nomatch = NULL;
+    obj_op.meta.user_data = NULL;
+    obj_op.meta.zones_trace = NULL;
+    obj_op.meta.delete_at = real_time();
+    obj_op.meta.olh_epoch = tier_ctx.o.versioned_epoch;
+    
+    RGWObjManifest *pmanifest; 
+
+    pmanifest = &(*s->manifest);
+    RGWObjTier tier_config;
+    tier_config.name = oc.tier.storage_class;
+    tier_config.tier_placement = oc.tier;
+    tier_config.is_multipart_upload = tier_ctx.is_multipart_upload;
+
+    pmanifest->set_tier_type("cloud");
+    pmanifest->set_tier_config(tier_config);
+
+    /* check if its necessary */
+    rgw_placement_rule target_placement;
+    target_placement.inherit_from(tier_ctx.bucket_info.placement_rule);
+    target_placement.storage_class = oc.tier.storage_class;
+    pmanifest->set_head(target_placement, tier_ctx.obj, 0);
+
+    pmanifest->set_tail_placement(target_placement, tier_ctx.obj.bucket);
+
+    /* should the obj_size also be set to '0' or is it needed
+     * to keep track of original size before transition. 
+     * But unless obj_size is set to '0', obj_iters cannot
+     * be reset I guess. For regular transitioned objects
+     * obj_size remains the same even when object is moved to other
+     * storage class. So maybe better to keep it the same way.
+     */
+    //pmanifest->set_obj_size(0);
+
+    obj_op.meta.manifest = pmanifest;
+
+    /* update storage class */
+    bufferlist bl;
+    bl.append(oc.tier.storage_class);
+    attrs[RGW_ATTR_STORAGE_CLASS] = bl;
+
+    attrs.erase(RGW_ATTR_ID_TAG);
+    attrs.erase(RGW_ATTR_TAIL_TAG);
+
+    obj_op.write_meta(tier_ctx.o.meta.size, 0, attrs, null_yield);
+    if (r < 0) {
+      return r;
+    }
+
+    return 0;
+  }
+
+  int transition_obj_to_cloud(lc_op_ctx& oc) {
+    std::shared_ptr<RGWRESTConn> conn;
+
+    /* init */
+    string id = "cloudid";
+    string endpoint=oc.tier.endpoint; 
+    RGWAccessKey key = oc.tier.key;
+    HostStyle host_style = oc.tier.host_style;
+    string bucket_name = oc.tier.target_path;
+    const RGWZoneGroup& zonegroup = oc.store->svc()->zone->get_zonegroup();
+   
+    if (bucket_name.empty()) {
+      bucket_name = "rgwx-" + zonegroup.get_name() + "-" + oc.tier.storage_class +
+                    "-cloud-bucket";
+      boost::algorithm::to_lower(bucket_name);
+    }
+
+    conn.reset(new S3RESTConn(oc.cct, oc.store->svc()->zone,
+                                id, { endpoint }, key, host_style));
+
+    /* http_mngr */
+    RGWCoroutinesManager crs(oc.store->ctx(), oc.store->getRados()->get_cr_registry());
+    RGWHTTPManager http_manager(oc.store->ctx(), crs.get_completion_mgr());
+
+    int ret = http_manager.start();
+    if (ret < 0) {
+      ldpp_dout(oc.dpp, 0) << "failed in http_manager.start() ret=" << ret << dendl;
+      return ret;
+    }
+
```

Comment:
I'm pretty sure there is;  can't we keep a timed cache of them around somewhere, linked from RGWLC?

Reply:
agreed that we'll want to address this, especially because `RGWHTTPManager::start()` is spawning a background thread each time. it would be great if we could just use the global `RGWHTTPManager`, but we need one that's coupled to the `RGWCompletionManager` of this `RGWCoroutinesManager`

Reply:
Okay.. To address this, created a global http_manager  & crs (specific to
RGWLC). This http_manager & crs will be initialized and started  (if not already done) the first time an entry of a bucket is
being transitioned to the cloud . To avoid keeping them active forever, they are released post processing all the lc entries in that bucket.

Please let me know if this sounds fine.